// @momsfriendlydevco/express-middleware-formatter/formats/html.js
var _ = require('lodash');
var xlsx = require('xlsx');

module.exports = {
	id: 'html',
	settings: {
		html: {
			download: false,
			filename: 'Exported Data.html',
			passthru: false,
			header: '<html><title>Exported data</title><body>',
			footer: '</body></html>',
		},
	},
	transform: function(emf, settings, content, req, res, next) {
		if (!_.isArray(content)) return next('Data is not suitable for the CSV output format');

		var workbook = xlsx.utils.book_new();
		var worksheet = xlsx.utils.json_to_sheet(content.map(i => emf.flatten(i)));
		xlsx.utils.book_append_sheet(workbook, worksheet, settings.xlsx.sheetName);
		var outputBuffer = xlsx.write(workbook, {
			type: 'buffer',
			bookType: 'html',
			header: settings.html.header,
			footer: settings.html.footer,
		});

		if (!settings.html.passthru) {
			res.type('html');
			if (settings.html.download) res.set('Content-Disposition', `attachment; filename="${settings.filename ? settings.filename + '.html' : settings.html.filename}"`);
			res.send(outputBuffer);
			next('STOP');
		} else {
			next(null, outputBuffer);
		}
	},
};
